test/standalone/test_rpc_proxy.py

```diff
@@ -410,3 +413,121 @@ def __setitem__(self, key, value):
     for _, message in fake_replies.messages:
         consumer_tags.add(message.delivery_info['consumer_tag'])
     assert len(consumer_tags) == 1
+
+
+@skip_if_no_toxiproxy
+class TestStandaloneProxyDisconnections(object):
+
+    @pytest.fixture(autouse=True)
+    def container(self, container_factory, rabbit_config):
+
+        class Service(object):
+            name = "service"
+
+            @rpc
+            def echo(self, arg):
+                return arg
+
+        config = rabbit_config
+
+        container = container_factory(Service, config)
+        container.start()
+
+    @pytest.yield_fixture(autouse=True)
+    def disable_retry(self):
+        # we don't want to wait for multiple retry attempts before
+        # connection failures bubble out
+        with patch.object(MethodProxy, 'retry', new=False):
+            yield
+
+    @pytest.yield_fixture(autouse=True, params=[True, False])
+    def use_confirms(self, request):
+        with patch.object(MethodProxy, 'use_confirms', new=request.param):
+            yield request.param
+
+    @pytest.yield_fixture
+    def service_rpc(self, toxiproxy, rabbit_config):
+
+        config = rabbit_config
+        config['AMQP_URI'] = toxiproxy.uri
+
+        with ServiceRpcProxy("service", config) as proxy:
+            yield proxy
+
+    def test_normal(self, service_rpc):
+        assert service_rpc.echo(1) == 1
+        assert service_rpc.echo(2) == 2
+
+    def test_down(self, service_rpc, toxiproxy):
+        toxiproxy.disable()
+
+        with pytest.raises(socket.error) as exc_info:
+            service_rpc.echo(1)
+        assert "ECONNREFUSED" in str(exc_info.value)
+
+    def test_timeout(self, service_rpc, toxiproxy):
+        toxiproxy.timeout()
+
+        with pytest.raises(IOError) as exc_info:
+            service_rpc.echo(1)
+        assert "Socket closed" in str(exc_info.value)
+
+    def test_recover_from_down(self, service_rpc, toxiproxy, use_confirms):
+        if not use_confirms:
+            pytest.skip(
+                "unconfirmed messages will be lost after disconnection"
+            )
+
+        assert service_rpc.echo(1) == 1
+
+        toxiproxy.disable()
+
+        # publisher cannot connect, raises
+        with pytest.raises(IOError) as exc_info:
+            service_rpc.echo(2)
+        assert (
+            # expect the write to raise a BrokenPipe or, if it succeeds,
+            # the socket to be closed on the subsequent confirmation read
+            "Broken pipe" in str(exc_info.value) or
+            "Socket closed" in str(exc_info.value)
+        )
+
+        toxiproxy.enable()
+
+        # consumer discovers its socket was closed
+        # TODO: can we not reestablish the connection earlier?
```

Comment:
i think this is the next chance the proxy has to run code. also, note that you caught the `IOError`, which we wouldn't expect a regular caller to do, right?